File: src/suikailauncher/core/account/xbox.py
from suikailauncher.core.base import network,exceptions
from suikailauncher.core.base.logger import logger
import json


async def login_xbox_live(access_token:str) -> tuple[str,str]:
    login_data = {
        "Properties": {
            "AuthMethod": "RPS",
            "SiteName": "user.auth.xboxlive.com",
            "RpsTicket": access_token
        },
        "RelyingParty": "http://auth.xboxlive.com",
        "TokenType": "JWT"
    }
    
    user_auth_resp = await network.network_request("https://user.auth.xboxlive.com/user/authenticate","POST",data=json.dumps(login_data))
    if user_auth_resp.is_error():
        raise exceptions.LoginException("远程服务器返回错误",user_auth_resp.decode())
    user_auth_result = user_auth_resp.json()
    token,uhs=user_auth_result.get("Token"),user_auth_result.get("DisplayClaims").get("xui")[0].get("uhs")
    return token,uhs
async def login_xsts(access_token:str) -> tuple[str,str]:
    login_data = {
        "Properties": {
            "SandboxId": "RETAIL",
            "UserTokens": [
                access_token
            ]
        },
        "RelyingParty": "rp://api.minecraftservices.com/",
        "TokenType": "JWT"
    }
    user_auth_resp = await network.network_request("https://xsts.auth.xboxlive.com/xsts/authorize","POST",data=json.dumps(login_data))
    if user_auth_resp.is_error():
        raise exceptions.LoginException("远程服务器返回错误",user_auth_resp.decode())
    user_auth_result = user_auth_resp.json()
    token,uhs = user_auth_result.get("Token"),user_auth_result.get("DisplayClaims").get("xui")[0].get("uhs")
    return token,uhs

File: SuikaiLauncher.Core/Modules/Account/Xbox.cs
#pragma warning disable CS8619

using System.Text.Json.Nodes;
using SuikaiLauncher.Core.Base;
using SuikaiLauncher.Core.Override;

namespace SuikaiLauncher.Core.Account{
    public class XboxClient{
        internal async static Task<Tuple<bool,string,string>> XboxLiveAuthorize(string AccessToken){
            JsonObject ReqData = new(){
                ["Properties"] = new JsonObject(){
                    ["AuthMethod"] = "RPS",
                    ["SiteName"] = "user.auth.xboxlive.com",
                    ["RpsTicket"] = AccessToken
                },
                ["RelyingParty"] = "http://auth.xboxlive.com",
                ["TokenType"] = "JWT"
            };
            HttpResponseMessage UserAuthResult = await Network.NetworkRequest(
                "https://user.auth.xboxlive.com/user/authenticate",
                new Dictionary<string, string>(){
                    ["Content-Type"] = "application/json"
                },ReqData.ToJsonString(),
                method:"POST"
            );
            JsonNode Result = Json.GetJson((await FileIO.ReadAsString(UserAuthResult)));
            string? Token = Result["Token"]?.ToString();
            JsonNode? DisplayClaims = Result["DisplayClaims"];
            if (DisplayClaims is not null){
                JsonNode? Xui = DisplayClaims["xui"];
                if (Xui is not null) {
                    if(Xui.AsArray()[0] is not null){
                        string? XboxUserHash = Xui.AsArray()[0]?["uhs"]?.ToString();
                        if(Token.IsNullOrWhiteSpaceF() || XboxUserHash.IsNullOrWhiteSpaceF()) return Tuple.Create(true,Token,XboxUserHash);
                    }
                }
            }
            return Tuple.Create(false,string.Empty,string.Empty);
        }
        internal async static Task<Tuple<bool,string?>> XSTSAuthorize(string AccessToken){
            JsonObject ReqData = new(){
                ["Properties"] = new JsonObject(){
                    ["SandboxId"] = "RETAIL",
                    ["UserTokens"] = new JsonArray(){
                        AccessToken
                    }
                },
                ["RelyingParty"] = "rp://api.minecraftservices.com",
                ["TokenType"] = "JWT"
            };
            HttpResponseMessage UserAuthResult = await Network.NetworkRequest(
                "https://xsts.auth.xboxlive.com/xsts/authorize",
                new Dictionary<string, string>(){
                    ["Content-Type"] = "application/json"
                },
                ReqData.ToJsonString(),
                method:"POST"
                );
            JsonNode Result = Json.GetJson(await FileIO.ReadAsString(UserAuthResult));
            string? Token = Result["Token"]?.ToString();
            if (Token.IsNullOrWhiteSpaceF()) return Tuple.Create(true,Token);
            return Tuple.Create<bool,string?>(false,string.Empty);
        }
    }
}